### PR TITLE
Add retry settings configuration to GCS plugin (max_retries)

### PR DESCRIPTION
**Pull request by reta**
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

Add retry settings configuration to GCS plugin (max_retries), borrowing other configuration parameters from the defaults. Added `GcsRetryableReadChannel` to resume reading partially downloaded blobs.

- [X] Add tests

### CODE REVIEW

File: repository-gcs/src/main/java/io/aiven/elasticsearch/repositories/gcs/GcsSettingsProvider.java [hunk 1]
```diff
@@ -73,6 +73,13 @@ private Storage createGcsClient(final GcsStorageSettings gcsStorageSettings) {
                                 .setReadTimeout(gcsStorageSettings.readTimeout())
                                 .build())
                 .setHeaderProvider(() -> Map.of(HttpHeaders.USER_AGENT, HTTP_USER_AGENT))
+                .setRetrySettings(
+                        // Create retry settings from default by modifying the max attempts settings
+                        StorageOptions
+                            .getDefaultRetrySettings()
```
Comment (reta) [inline repository-gcs/src/main/java/io/aiven/elasticsearch/repositories/gcs/GcsSettingsProvider.java:79]: Here are the defaults:

```
    return RetrySettings.newBuilder()
        .setMaxAttempts(6)
        .setInitialRetryDelay(Duration.ofMillis(1000L))
        .setMaxRetryDelay(Duration.ofMillis(32_000L))
        .setRetryDelayMultiplier(2.0)
        .setTotalTimeout(Duration.ofMillis(50_000L))
        .setInitialRpcTimeout(Duration.ofMillis(50_000L))
        .setRpcTimeoutMultiplier(1.0)
        .setMaxRpcTimeout(Duration.ofMillis(50_000L));
```

Do we see a value in providing the configuration settings for all / some of them, along with `max_retries`?

File: checkstyle/suppressions.xml [hunk 2]
```diff
@@ -19,4 +19,5 @@
         "https://checkstyle.org/dtds/suppressions_1_2.dtd">
 
 <suppressions>
+    <suppress id="maxParameterNumber" files="GcsStorageSettings.java"/>
```
Comment (reta) [inline checkstyle/suppressions.xml:22]: Goes a bit over 10 parameters limit, probably fine for the settings holder.

File: repository-gcs/src/main/java/io/aiven/elasticsearch/repositories/gcs/GcsRetryableReadChannel.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+/**
+ * Wrapper around reads from GCS that will retry blob downloads that fail part-way through, resuming from where the 
+ * failure occurred. It uses the fact that ReadChannel supports capture and restore functionality and could be recreated
+ * from the last successful checkpoint.
+ */
+public class GcsRetryableReadChannel implements ReadableByteChannel {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcsRetryableReadChannel.class);
+    private static final int MAX_SUPPRESSED_EXCEPTIONS = 10;
+
+    private volatile ReadChannel delegate;
```
Comment (willyborankin) [inline repository-gcs/src/main/java/io/aiven/elasticsearch/repositories/gcs/GcsRetryableReadChannel.java:45]: No need to use `volatile` `GcsRetryableReadChannel` is thread safe. It needs to be `final`. :-)

File: repository-gcs/src/test/java/io/aiven/elasticsearch/repositories/gcs/GcsRetryableReadChannelTest.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class GcsRetryableReadChannelTest {
```
[the remark this answers is not in the document]
Comment (reta) [reply repository-gcs/src/test/java/io/aiven/elasticsearch/repositories/gcs/GcsRetryableReadChannelTest.java:50]: :+1:

File: repository-gcs/src/main/java/io/aiven/elasticsearch/repositories/gcs/GcsStorageSettings.java [hunk 5]
```diff
@@ -67,6 +67,11 @@
             Setting.intSetting(withPrefix("gcs.client.read_timeout"), -1, -1,
                     Setting.Property.NodeScope);
 
+    /** The number of retries to use when an GCS request fails. */
+    public static final Setting<Integer> MAX_RETRIES_SETTING =
```
[the remark this answers is not in the document]
Comment (reta) [reply repository-gcs/src/main/java/io/aiven/elasticsearch/repositories/gcs/GcsStorageSettings.java:71]: :+1:

**Pull request merged**